File: backend/schemas/simulation.py
from pydantic import BaseModel, Field


class SimulationRequest(BaseModel):
    max_order: int = Field(default=3, ge=1, le=8)
    rir_duration: float = Field(default=1.0, ge=0.2, le=5.0)


class SimulationResult(BaseModel):
    rir_id: str
    rt60_sabine: float
    rt60_computed: float | None
    direct_delay_ms: float
    reflection_count: int
    job_id: str


class JobStatus(BaseModel):
    job_id: str
    status: str  # pending | running | done | error
    progress: float = 0.0
    result: dict | None = None
    error: str | None = None

File: frontend/src/api/simulation.ts
import { apiClient } from "./client";

export async function startSimulation(
  sessionId: string,
  maxOrder = 6,
  rirDuration = 1.0,
) {
  const { data } = await apiClient.post(`/sessions/${sessionId}/simulate`, {
    max_order: maxOrder,
    rir_duration: rirDuration,
  });
  return data as { job_id: string; status: string };
}

export async function simulatePreview(sessionId: string) {
  const { data } = await apiClient.post(`/sessions/${sessionId}/simulate/preview`);
  return data as {
    rt60_sabine: number;
    rt60_computed: number | null;
    direct_delay_ms: number;
    reflection_count: number;
  };
}

export async function getSimulationStatus(sessionId: string, jobId: string) {
  const { data } = await apiClient.get(
    `/sessions/${sessionId}/simulate/status/${jobId}`,
  );
  return data as {
    job_id: string;
    status: string;
    result?: Record<string, unknown>;
    error?: string;
  };
}

export function createProgressEventSource(sessionId: string, jobId: string): EventSource {
  return new EventSource(`/api/sessions/${sessionId}/simulate/progress/${jobId}`);
}
